src/main/java/opendcs/dao/ScheduleEntryDAO.java

```diff
@@ -56,489 +56,596 @@
-				if (!se.getLoadingAppId().isNull())
-					for(CompAppInfo appInfo : appInfos)
-						if (se.getLoadingAppId().equals(appInfo.getAppId()))
-						{
-							se.setLoadingAppName(appInfo.getAppName());
-							break;
-						}
-				ret.add(se);
+                throw new NoSuchObjectException("ScheduleEntry id="
+                    + scheduleEntry.getKey() + " '"
+                    + scheduleEntry.getName() + "' does not exist in database.");
 			}
+        }
+        catch(SQLException ex)
+        {
+            throw new DbIoException("Error checking ScheduleEntry id="
+                + scheduleEntry.getKey() + " '"
+                + scheduleEntry.getName() + "': ", ex);
+        }
+    }
+
+    @Override
+    public void writeScheduleEntry(ScheduleEntry scheduleEntry) throws DbIoException
+    {
+        if (db.getDecodesDatabaseVersion() < DecodesDatabaseVersion.DECODES_DB_10)
+		{
+            return;
+		}
+
+		debug3("writeScheduleEntry(" + scheduleEntry.getName() + ") rsID=" + scheduleEntry.getRoutingSpecId()
+				+ ", rsname='" + scheduleEntry.getRoutingSpecName() + "', appID=" + scheduleEntry.getLoadingAppId()
+				+ ", appName='" + scheduleEntry.getLoadingAppName() + "'");
+        if (scheduleEntry.getRoutingSpecId().isNull()
+         && scheduleEntry.getRoutingSpecName() != null
+         && scheduleEntry.getRoutingSpecName().length() > 0)
+        {
+            String q = "select id from RoutingSpec where upper(name) = upper(?)";
+            try
+            {
+				doQuery(q, rs -> scheduleEntry.setRoutingSpecId(DbKey.createDbKey(rs, 1)), scheduleEntry.getRoutingSpecName());
+            }
+            catch (SQLException ex)
+            {
+                throw new DbIoException("writeScheduleEntry Error in query '" + q + "'", ex);
+            }
+        }
+
+        if (scheduleEntry.getLoadingAppId().isNull()
+         && scheduleEntry.getLoadingAppName() != null
+         && scheduleEntry.getLoadingAppName().length() > 0)
+        {
+            String q = "select loading_application_id from "
+                + "hdb_loading_application where upper(loading_application_name) = upper(?)";
+            try
+            {
+				doQuery(q, rs -> scheduleEntry.setLoadingAppId(DbKey.createDbKey(rs, 1)), scheduleEntry.getLoadingAppName());
+            }
+            catch (SQLException ex)
+            {
+                throw new DbIoException("writeScheduleEntry Error in query '" + q + "'", ex);
+            }
+        }
+
+
+        scheduleEntry.setLastModified(new Date());
+        // It might be an import from an xml file. If no key, try to lookup from name.
+        if (scheduleEntry.getKey().isNull())
+        {
+            String q = "select schedule_entry_id from schedule_entry where "
+                + " upper(name) = upper(?)";// + sqlString(scheduleEntry.getName().toUpperCase());
+            try
+            {
+				doQuery(q, rs -> scheduleEntry.forceSetId(DbKey.createDbKey(rs, 1)), scheduleEntry.getName());
```

Comment:
The `ScheduledEntry` provided to this function should have a name already but it may be from a reference that wasn't loaded from the database so this looks up the database id for that name. and then assigns the id.

This is then used to determine if the entry is an update or new entry below. Might be a good idea to put check for a valid name at the top of the method though just to make that clear.